src/libraries/component-scene2d/Scene2d.tsx

```diff
@@ -0,0 +1,312 @@
+import { BaseCanvas, pointInRect, RectangularRegion, Vec4 } from "libraries/figurl-canvas";
+import { dragSelectReducer } from "libraries/util-drag-select";
+import React, { FunctionComponent, useCallback, useEffect, useReducer, useState } from "react";
+
+// https://figurl.org/f?v=gs://figurl/spikesortingview-9&d=sha1://309f3b0131934091301c3e75669138a2e04e9240&label=Test1
+// https://figurl.org/f?v=http://localhost:3000&d=sha1://309f3b0131934091301c3e75669138a2e04e9240&label=Test1
+
+type T = {
+	type: 'line'
+	x: number
+	y: number
+	dx: number
+	dy: number
+	attributes: {
+		color: string
+		dash?: number[]
+		width?: number
+	}
+	selectedAttributes?: {
+		color: string
+		dash?: number[]
+		width?: number
+	}
+} | {
+	type: 'marker'
+	x: number
+	y: number
+	attributes: {
+		fillColor?: string
+		lineColor?: string
+		shape?: 'circle' | 'square'
+		radius?: number
+	}
+	selectedAttributes?: {
+		fillColor?: string
+		lineColor?: string
+		shape?: 'circle' | 'square'
+		radius?: number
+	}
+} | {
+	type: 'connector',
+	objectId1: string,
+	objectId2: string,
+	attributes: {
+		color: string
+		dash?: number[]
+		width?: number
+	}
+}
+
+export type Scene2dObject = {
+	objectId: string
+	clickable?: boolean
+	draggable?: boolean
+	selected?: boolean
+} & T
+
+type Props ={
+	width: number
+	height: number
+	objects: Scene2dObject[]
+	onClickObject?: (objectId: string, e: React.MouseEvent) => void
+	onDragObject?: (objectId: string, newPoint: {x: number, y: number}, e: React.MouseEvent) => void
+	onSelectObjects?: (objectIds: string[], e: React.MouseEvent | undefined) => void
+	onClick?: (p: {x: number, y: number}, e: React.MouseEvent) => void
+}
+
+const emptyDrawData = {}
+
+const defaultMarkerRadius = 6
+const defaultLineWidth = 1.1
+
+type DraggingObjectState = {
+	object?: Scene2dObject | null
+	newPoint?: {x: number, y: number}
+}
+type DraggingObjectAction = {
+	type: 'start'
+	object: Scene2dObject | null
+	point: {x: number, y: number}
+} | {
+	type: 'end'
+} | {
+	type: 'move'
+	point: {x: number, y: number}
+}
+
+const draggingObjectReducer = (s: DraggingObjectState, a: DraggingObjectAction): DraggingObjectState => {
+	if (a.type === 'start') {
+		return {...s, object: a.object, newPoint: a.point}
+	}
+	else if (a.type === 'end') {
+		return {...s, object: undefined}
+	}
+	else if (a.type === 'move') {
+		return {...s, newPoint: a.point}
+	}
+	else return s
+}
+
+const Scene2d: FunctionComponent<Props> = ({width, height, objects, onClickObject, onDragObject, onSelectObjects, onClick}) => {
+	const [dragState, dragStateDispatch] = useReducer(dragSelectReducer, {})
+	const [draggingObject, draggingObjectDispatch] = useReducer(draggingObjectReducer, {})
+	const [activeSelectRect, setActiveSelectRect] = useState<Vec4 | undefined>()
+	const [activeMouseEvent, setActiveMouseEvent] = useState<React.MouseEvent | undefined>()
+
+	const handleSelectRect = useCallback((r: Vec4, e: React.MouseEvent | undefined) => {
```

Comment:
The `handleSelectRect`, subsequent `useEffect`, and `paint` logic are all pretty hairy--not going to fuss about it right now but I think breaking them out into smaller functions would make it easier to follow. I also wonder if more of the select-rect stuff should be provided by the `useDragSelect` library code.

Reply:
looking at the documented version might help (recent PR)